go/libkb/naclwrap.go

```diff
@@ -700,20 +708,61 @@ func (k NaclDHKeyPair) EncryptToString(plaintext []byte, sender GenericKey) (str
 	return PacketArmoredEncode(info)
 }
 
-func (k NaclDHKeyPair) SecretSymmetricKey(reason EncryptionReason) ([]byte, error) {
-
+func (k NaclDHKeyPair) SecretSymmetricKey(reason EncryptionReason) (NaclSecretBoxKey, error) {
 	if !k.CanDecrypt() {
-		return nil, NoSecretKeyError{}
+		return NaclSecretBoxKey{}, NoSecretKeyError{}
 	}
-	if len(reason) < 8 {
-		return nil, KeyGenError{Msg: "reason must be at least 8 bytes"}
+
+	return deriveSymmetricKeyFromAsymmetric(*k.Private, reason)
+}
+
+// Derive a symmetric key using HMAC(k, reason).
+// Suitable for deriving from an asymmetric encryption key.
+// For deriving from a shared encryption key, this output is too close
+// to something that might be used as a public authenticator.
+func deriveSymmetricKeyFromAsymmetric(inKey NaclDHKeyPrivate, reason EncryptionReason) (NaclSecretBoxKey, error) {
+	var outKey = [32]byte{}
+	if len(reason) < encryptionReasonMinLength {
+		return outKey, KeyGenError{Msg: "reason must be at least 8 bytes"}
+	}
+
+	mac := hmac.New(sha256.New, inKey[:])
+	_, err := mac.Write(reason.Bytes())
+	if err != nil {
+		return outKey, err
 	}
+	out := mac.Sum(nil)
 
-	mac := hmac.New(sha256.New, k.Private[:])
-	mac.Write(reason.Bytes())
-	symmetricKey := mac.Sum(nil)
+	if copy(outKey[:], out) != len(outKey) {
+		return outKey, KeyGenError{Msg: "derived key of wrong size"}
+	}
+
+	return outKey, nil
+}
+
+// Derive a symmetric key using HKDF.
+func DeriveSymmetricKey(inKey NaclSecretBoxKey, reason EncryptionReason) (NaclSecretBoxKey, error) {
+	if len(reason) < encryptionReasonMinLength {
+		return NaclSecretBoxKey{}, KeyGenError{Msg: "reason must be at least 8 bytes"}
+	}
+
+	hash := sha256.New
+	info := []byte(reason)
+
+	// Use nil salt
+	stream := hkdf.New(hash, inKey[:], nil, info)
```

Comment:
Like we talked about yesterday / last week, I'm less enthusiastic about HKDF than I was when we first started. Let's switch to...anything else :) Your suggestion of using the context string as an HMAC key, with the symmetric key as the input, sounds great.